### PR TITLE
gtk: add IMContext subclassing support

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Part of #56

### CODE REVIEW

File: gtk4/src/subclass/im_context.rs [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+    }
+    fn retrieve_surrounding(&self, im_context: &Self::Type) -> bool {
+        self.parent_retrieve_surrounding(im_context)
+    }
+    fn set_client_widget<P: IsA<Widget>>(&self, im_context: &Self::Type, widget: Option<&P>) {
+        self.parent_set_client_widget(im_context, widget)
+    }
+    fn set_cursor_location(&self, im_context: &Self::Type) -> Option<gdk::Rectangle> {
+        self.parent_set_cursor_location(im_context)
+    }
```
Comment (bilelmoussaoui) [inline gtk4/src/subclass/im_context.rs:50]: I'm not sure whether this method should return the area or not...

Comment (sdroege) [reply gtk4/src/subclass/im_context.rs:50]: No, it's the other way around. The caller passes a rectangle (not nullable!) in. See `gtk_im_context_set_cursor_location()` docs.

File: gtk4/src/subclass/im_context.rs [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    let area = imp.set_cursor_location(wrap.unsafe_cast_ref());
+    *areaptr = *area.to_glib_none().0;
+}
+
+unsafe extern "C" fn im_context_set_surrounding<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    textptr: *const libc::c_char,
+    _length: i32,
```
Comment (sdroege) [inline gtk4/src/subclass/im_context.rs:490]: You probably want to make use of this

Comment (bilelmoussaoui) [reply gtk4/src/subclass/im_context.rs:490]: Use it as in creating a string from textptr for only the length of bytes passed? that makes sense i guess

File: gtk4/src/subclass/im_context.rs [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+unsafe extern "C" fn im_context_set_surrounding<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    textptr: *const libc::c_char,
+    length: i32,
+    cursor_index: i32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let text: GString = FromGlibContainer::from_glib_none_num(textptr, length as usize);
```
Comment (bilelmoussaoui) [inline gtk4/src/subclass/im_context.rs:495]: The C code does uses `g_strndup (text, len);` to only get the passed length, does it make sense to a transfer: none here?

Comment (sdroege) [reply gtk4/src/subclass/im_context.rs:494]: What C code? And why not use `from_glib_borrow()` here?

Comment (bilelmoussaoui) [reply gtk4/src/subclass/im_context.rs:494]: But then I would have to convert the gstring to bytes to only take `length` bytes & convert it back to a gstring?

Comment (sdroege) [reply gtk4/src/subclass/im_context.rs:494]: Pff... hmm.

```rust
let s = GString::from_glib_borrow(textptr);
f(&s[..length]);
```

or something like that maybe? The `from_glib_none()` there does a full copy of the string :)

Comment (bilelmoussaoui) [reply gtk4/src/subclass/im_context.rs]: That worked fine :) thanks!

File: gtk4/src/subclass/im_context.rs [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+    fn reset(&self, im_context: &Self::Type) {
+        self.parent_reset(im_context)
+    }
+    fn retrieve_surrounding(&self, im_context: &Self::Type) -> bool {
+        self.parent_retrieve_surrounding(im_context)
+    }
+    fn set_client_widget<P: IsA<Widget>>(&self, im_context: &Self::Type, widget: Option<&P>) {
+        self.parent_set_client_widget(im_context, widget)
+    }
+    fn set_cursor_location(&self, im_context: &Self::Type, area: &mut gdk::Rectangle) {
```
Comment (sdroege) [inline gtk4/src/subclass/im_context.rs:48]:
```suggestion
    fn set_cursor_location(&self, im_context: &Self::Type, area: &gdk::Rectangle) {
```

etc according to the docs

File: gtk4/src/subclass/im_context.rs [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+}
+
+impl<T: IMContextImpl> IMContextImplExt for T {
+    fn parent_commit(&self, im_context: &Self::Type, string: &str) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut ffi::GtkIMContextClass;
+            let f = (*parent_class)
+                .commit
+                .expect("No parent class impl for \"commit\"");
```
Comment (sdroege) [inline gtk4/src/subclass/im_context.rs:86]: This is a signal. That's usually a good sign for a vfunc to be optional :)

File: gtk4/src/subclass/im_context.rs [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    let (text, attrs, cursor_idx) = imp.get_preedit_string(wrap.unsafe_cast_ref());
+
+    *text_ptr = text.to_glib_none().0;
+    *cursor_index_ptr = cursor_idx;
+    *attrs_ptr = attrs.to_glib_none().0;
```
Comment (sdroege) [inline gtk4/src/subclass/im_context.rs:398]: The docs say that these are all `transfer full`

File: gtk4/src/subclass/im_context.rs [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+    ptr: *mut ffi::GtkIMContext,
+    text_ptr: *mut *mut libc::c_char,
+    cursor_index_ptr: *mut libc::c_int,
+) -> glib::ffi::gboolean {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    if let Some((text, cursor_idx)) = imp.get_surrounding(wrap.unsafe_cast_ref()) {
+        *text_ptr = text.to_glib_none().0;
```
Comment (sdroege) [inline gtk4/src/subclass/im_context.rs:411]: `transfer full`

File: gtk4/src/subclass/im_context.rs [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+        *cursor_index_ptr = cursor_idx;
+        return true.to_glib();
+    }
+    false.to_glib()
```
Comment (sdroege) [inline gtk4/src/subclass/im_context.rs:415]: Maybe to be friendly you want to set `text_ptr` and `cursor_index_ptr` to `NULL` / `0` in this case

File: gtk4/src/subclass/im_context.rs [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+}
+
+unsafe extern "C" fn im_context_set_cursor_location<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    areaptr: *mut gdk::ffi::GdkRectangle,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let mut area = gdk::Rectangle::uninitialized();
```
Comment (sdroege) [inline gtk4/src/subclass/im_context.rs:479]: The caller provides the area rect and you just pass a read-only reference of it onwards

File: gtk4/src/subclass/im_context.rs [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+    cursor_index: i32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let text: Borrowed<GString> = from_glib_borrow(textptr);
+
+    imp.set_surrounding(
+        wrap.unsafe_cast_ref(),
+        &text[0..(length as usize)],
```
Comment (sdroege) [inline gtk4/src/subclass/im_context.rs:498]: > the length of text , or -1 if text is nul-terminated

So you need to handle `-1` here differently. `-1 == usize::MAX` would panic here now :)

**Review by sdroege — CHANGES_REQUESTED** [review]
Generally you have to compare things to the docs a bit more :)

**Review by sdroege — APPROVED** [review]

**Pull request merged**